File: 03.0-very-simple-http-server/server.3.valid-http-response.rb
require 'socket' # A library built into Ruby - provides the TCPServer class
require '../lib/unbuffered'

$stdout.sync = 1  # Line-buffer output to STDOUT.
listener = TCPServer.open(8080)
loop do
  socket = listener.accept  # Wait til a client connects, then open a socket.
  puts "FROM THE CLIENT: "
  txt = read_line_unbuffered_from(socket) # The first line, which is the Request-Line
  (method,uri,version) = txt.strip.split(' ')
  puts "METHOD = #{method}, URI = #{uri}, version = #{version}"

  while txt = read_line_unbuffered_from(socket)
    puts txt
    if txt == "\r\n"  # i.e., CRLF
      puts "SERVER: reached end of request headers"
      socket.print("HTTP/1.1 200 OK\r\n")
      socket.print("\r\n")
      socket.puts("Hello, world")
      puts "SERVER: Sent 'Hello, world' response to client"
      socket.close
      break
    end
  end
end
puts "SERVER: now I'm exiting"

File: 03.0-very-simple-http-server/server.3.valid-http-response.py
import socket
import os
import sys
from unbuffered_reading import read_line_unbuffered_from

sys.stdout = os.fdopen(sys.stdout.fileno(), 'w', 1)  # Line-buffer output to STDOUT.
listener = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
listener.bind(('0.0.0.0', 8080))
listener.listen(5)
while 1:
  try:
    (socket,address) = listener.accept()
    txt = read_line_unbuffered_from(socket).decode('utf8')
    print("FROM THE CLIENT: ")
    while txt != "\r\n":  # i.e., CRLF
      print(txt)
      txt = read_line_unbuffered_from(socket).decode('utf8')
    print("SERVER: reached end of request headers")
    socket.send("HTTP/1.1 200 OK\r\n".encode('utf8'))
    socket.send("\r\n".encode('utf8'))
    socket.send("Hello, world\n".encode('utf8'))
    print("SERVER: Sent 'Hello, world' response to client")
    socket.close()
  except KeyboardInterrupt:
    print("Caught Ctrl-C, exiting...")
    listener.close()
    exit(1)

print("SERVER: now I'm exiting")
